ori/volume/server/volume_create.go

```diff
@@ -16,261 +16,143 @@ package server
-	apiutils.SetVolumeClassLabel(pvc, volume.Spec.Class)
-
-	return &AggregateCephVolume{
-		Pvc:         pvc,
-		ImagePvc:    imagePvc,
-		VolumeClass: volumeClass,
-	}, nil
-}
 
-func (s *Server) handleImagePopulation(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) error {
-	if volume.ImagePvc == nil {
-		return nil
-	}
-
-	snapshot := &snapshotv1.VolumeSnapshot{}
-	if err := s.client.Get(ctx, client.ObjectKeyFromObject(volume.ImagePvc), snapshot); err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("unable to get snapshot: %w", err)
+		imageSize := uint64(onmetalImage.RootFS.Descriptor().Size)
+		if volume.Spec.Resources.StorageBytes <= imageSize {
+			return nil, fmt.Errorf("volume size  (%d) smaller than image (%d)", volume.Spec.Resources.StorageBytes, imageSize)
 		}
-		log.V(1).Info("Requested snapshot not found: create image pvc and snapshot it.")
-		return s.createSnapshot(ctx, log, volume)
-	}
-
-	if !pointer.BoolDeref(snapshot.Status.ReadyToUse, false) || snapshot.Status.RestoreSize == nil {
-		log.Info("Referenced snapshot not ready or RestoreSize not defined", "snapshotName", snapshot.Name)
-		return nil
-	}
-
-	volumeSizeBytes := volume.Pvc.Spec.Resources.Requests.Storage().Value()
-	if volumeSizeBytes < snapshot.Status.RestoreSize.Value() {
-		log.Info(fmt.Sprintf("Requested volume size %d is less than the size %d for the source snapshot", volumeSizeBytes, snapshot.Status.RestoreSize.Value()), "snapshotName", snapshot.Name)
-		//r.Eventf(volume, corev1.EventTypeWarning, ReasonVolumeSizeToSmall, "Requested volume size %d is less than the size %d for the source snapshot %s", volumeSizeBytes, snapshot.Status.RestoreSize.Value(), snapshot.Name)
-	}
 
-	return nil
-}
-
-func (s *Server) createSnapshot(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) error {
-	if err := s.client.Create(ctx, volume.ImagePvc); err != nil {
-		return fmt.Errorf("error creating pvc: %w", err)
+		snapshotRef = pointer.String(img.Descriptor().Digest.String())
+		snapshot = &api.Snapshot{
+			Metadata: api.Metadata{
+				ID: img.Descriptor().Digest.String(),
+			},
+			Source: api.SnapshotSource{
+				OnmetalImage: image,
+			},
+		}
+		labels[SnapshotNameLabel] = image
 	}
-	log.V(1).Info(fmt.Sprintf("created image pvc %s", client.ObjectKeyFromObject(volume.ImagePvc)))
 
-	snapshot := &snapshotv1.VolumeSnapshot{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "VolumeSnapshot",
-			APIVersion: snapshotv1.SchemeGroupVersion.String(),
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      volume.ImagePvc.Name,
-			Namespace: volume.ImagePvc.Namespace,
-		},
-		Spec: snapshotv1.VolumeSnapshotSpec{
-			Source: snapshotv1.VolumeSnapshotSource{
-				PersistentVolumeClaimName: &volume.ImagePvc.Name,
+	return &aggregatedImage{
+		image: &api.Image{
+			Metadata: api.Metadata{
+				ID:          s.idGen.Generate(),
+				Annotations: volume.Metadata.Annotations,
+				Labels:      labels,
```

Comment:
There has to be a separation between labels & annotations required by the `ori` server implementation and labels and annotations specified by the `CreateVolume` call.